### PR TITLE
GitHub Issue template

### LINKED ISSUE #1407: Introduce GitHub issue template
We can introduce GitHub issue template.

For example, CPython has already introduced this feature. https://github.com/python/cpython/tree/main/.github/ISSUE_TEMPLATE

We can have templates for questions, bug reports, feature requests, etc.

The following are the screenshots from CPython GitHub issue pages:

image

image

### PR DESCRIPTION
**Pull request by tigercosmos**
Added GitHub issue template. fix #1407

The demos are as follows:

image

image

### CODE REVIEW

File: .github/ISSUE_TEMPLATE/bug.yml [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+        - "PcapPlusPlus master branch"
+        - "v23.09"
+        - "v22.11"
+        - "v22.05"
+        - "v21.11"
+        - "v21.05"
+        - "Other"
+    validations:
+      required: true
+  - type: dropdown
```
Comment (egecetin) [inline .github/ISSUE_TEMPLATE/bug.yml:49]: Maybe we can add a field for compiler type 🤔 (MSVC, mingw, GNU, Intel ...) and maybe a comment like `Please also specify compiler version`. Sometimes newest versions breaks something like recent fail in CI (warning to error)

File: .github/ISSUE_TEMPLATE/bug.yml [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+      label: "Operating systems tested on:"
+      multiple: true
+      options:
+        - Linux
+        - macOS
+        - Windows
+        - FreeBSD
+        - Android
+        - Other
+    validations:
```
Comment (egecetin) [inline .github/ISSUE_TEMPLATE/bug.yml:60]: We can also add `Packet capture backend (if applicable)` (DPDK, PF_RING, libpcap, winpcap, npcap, N/A ...)

File: .github/ISSUE_TEMPLATE/bug.yml [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    attributes:
+      label: "Operating systems tested on:"
+      multiple: true
+      options:
+        - Linux
+        - macOS
+        - Windows
+        - FreeBSD
+        - Android
+        - Other
```
Comment (Dimi1010) [inline .github/ISSUE_TEMPLATE/bug.yml:59]: Not really familiar with what the template yaml supports, but maybe we can add a text box that shows when "Other" is selected, so the issuer can specify the OS in question, if conditional hiding of fields is supported?

Comment (tigercosmos) [reply .github/ISSUE_TEMPLATE/bug.yml:66]: I added an additional "input" for "Other".

File: .github/ISSUE_TEMPLATE/bug.yml (hunk 2 above)
Comment (tigercosmos) [reply .github/ISSUE_TEMPLATE/bug.yml:67]: added.

File: .github/ISSUE_TEMPLATE/bug.yml (hunk 1 above)
Comment (tigercosmos) [reply .github/ISSUE_TEMPLATE/bug.yml:56]: added

File: .github/ISSUE_TEMPLATE/bug.yml [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+      label: Compiler version:
+      description: Please specify the compiler version you used.
+      placeholder: ex. GCC 10.2.0
+    validations:
+      required: true
+  - type: dropdown
+    attributes:
+      label: Packet capture backend (if applicable):
+      multiple: true
+      options:
```
Comment (seladb) [inline .github/ISSUE_TEMPLATE/bug.yml:87]: We can add XDP as well

File: .github/ISSUE_TEMPLATE/question.yml [hunk 5]
```diff
@@ -0,0 +1,24 @@
+name: Question
```
Comment (seladb) [inline .github/ISSUE_TEMPLATE/question.yml:1]: I'd add a platform option here as well (could be optional). In many cases questions are for specific platforms

File: .github/ISSUE_TEMPLATE/bug.yml [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+      description: >
+        Give a clear and concise description of what happened.
+        Include a [minimal, reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) if possible.
+        [Copy and paste code where possible rather than using screenshots](https://meta.stackoverflow.com/a/285557/13990016),
+        and put any code blocks inside triple backticks.
+
+      value: |
+        ```cpp
+        # Add a code block here, if required
+        ```
```
Comment (seladb) [inline .github/ISSUE_TEMPLATE/bug.yml:32]: Should we add a bug report template with specific questions people need to answer to make the bug report more structured?

Comment (tigercosmos) [reply .github/ISSUE_TEMPLATE/bug.yml:32]: I think we already cover most of the questions that we are interested in, e.g. version, compiler, OS, etc.

Perhaps we can also add a section for backtrace, or error log, but I think these don't have to be a standalone question.

People should be able to add more information in the "description" section.

Do you have specific questions want to add to the template?

File: .github/ISSUE_TEMPLATE/question.yml (hunk 5 above)
Comment (tigercosmos) [reply .github/ISSUE_TEMPLATE/question.yml:1]: added

File: .github/ISSUE_TEMPLATE/bug.yml (hunk 4 above)
Comment (tigercosmos) [reply .github/ISSUE_TEMPLATE/bug.yml:87]: added

File: .github/ISSUE_TEMPLATE/bug.yml (hunk 6 above)
Comment (seladb) [reply .github/ISSUE_TEMPLATE/bug.yml:32]: Currently the bug description only includes a template for C++ code:
image

When I first looked at it I thought I only need to fill out the code example, and I searched where can I write the bug description 🙂 

I just opened a bug in Npcap and their template is quite nice, although we need different topics:
https://github.com/nmap/npcap/issues/new?assignees=&labels=&projects=&template=bug_report.md&title=

Maybe:
```markdown
**Describe the bug**
A clear and concise description of what the bug is.

**Code example to reproduce**
(code template)

**Expected behavior**
A clear and concise description of what you expected to happen.
```

Comment (tigercosmos) [reply .github/ISSUE_TEMPLATE/bug.yml:32]: Okay, I will apply your suggestion, but I would prevent putting too many words inside, most of the time, I also ignore those "templates" when I submit issues in other projects. If we put too much stuff, people will just feel annoyed and ignore it.

Comment (tigercosmos) [reply .github/ISSUE_TEMPLATE/bug.yml]: updated.

**Comment by tigercosmos** [comment]
@seladb  I guess this is ready for merge.

**Comment by seladb** [comment]
@clementperon @egecetin @Dimi1010 let me know when this PR is LGTM

I'll also take a look

File: .github/ISSUE_TEMPLATE/black.md [hunk 7]
```diff
@@ -0,0 +1,6 @@
+---
```
Comment (egecetin) [inline .github/ISSUE_TEMPLATE/black.md:1]: Filename should `black.md` or `blank.md`? Maybe a typo

**Review by egecetin — APPROVED** [review]
Only left one comment. I'm not sure is a typo or not.

File: .github/ISSUE_TEMPLATE/black.md (hunk 7 above)
Comment (tigercosmos) [reply .github/ISSUE_TEMPLATE/black.md:1]: ops, thanks!!

**Review by egecetin — APPROVED** [review]

**Review by Dimi1010 — APPROVED** [review]
LGTM

**Comment by seladb** [comment]
@tigercosmos can you merge https://github.com/tigercosmos/PcapPlusPlus/tree/issue_template to https://github.com/tigercosmos/PcapPlusPlus/tree/master so it'll be easier to test the different issues types?

File: .github/ISSUE_TEMPLATE/config.yml [hunk 8]
```diff
@@ -0,0 +1,7 @@
+contact_links:
+  - name: "Contact the PcapPlusPlus team"
+    about: "For help or advice on using PcapPlusPlus, try one of the following options instead of opening a GitHub issue:"
+    url: "https://pcapplusplus.github.io/community"
+  - name: "Documentation"
+    about: "Check the PcapPlusPlus documentation for more information."
+    url: "https://pcapplusplus.github.io/docs/quickstart"
```
Comment (seladb) [inline .github/ISSUE_TEMPLATE/config.yml:7]: Maybe this should redirect to https://pcapplusplus.github.io/ ?

**Comment by tigercosmos** [comment]
> @tigercosmos can you merge https://github.com/tigercosmos/PcapPlusPlus/tree/issue_template to https://github.com/tigercosmos/PcapPlusPlus/tree/master so it'll be easier to test the different issues types?

done, you can check now

File: .github/ISSUE_TEMPLATE/config.yml (hunk 8 above)
Comment (tigercosmos) [reply .github/ISSUE_TEMPLATE/config.yml:7]: sure, fixed.

**Comment by Dimi1010** [comment]
> > @tigercosmos can you merge https://github.com/tigercosmos/PcapPlusPlus/tree/issue_template to https://github.com/tigercosmos/PcapPlusPlus/tree/master so it'll be easier to test the different issues types?
> 
> done, you can check now

The bug report template has an issue
image

File: .github/ISSUE_TEMPLATE/bug.yml [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+  - type: input
+    attributes:
+      label: Compiler version:
+      description: Please specify the compiler version you used.
+      placeholder: ex. GCC 10.2.0
+    validations:
+      required: true
+  - type: dropdown
+    attributes:
+      label: Packet capture backend (if applicable):
```
Comment (Dimi1010) [inline .github/ISSUE_TEMPLATE/bug.yml:93]: Needs to be in "" because of the `:`

Comment (tigercosmos) [reply .github/ISSUE_TEMPLATE/bug.yml:93]: let me fix

Comment (tigercosmos) [reply .github/ISSUE_TEMPLATE/bug.yml]: @seladb  @Dimi1010  fixed.

File: .github/ISSUE_TEMPLATE/question.yml [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+
+  - type: dropdown
+    attributes:
+      label: "Operating systems:"
+      description: >
+        Select the operating systems that you are interested in.
+      multiple: true
+      options:
+        - Linux
+        - macOS
```
Comment (clementperon) [inline .github/ISSUE_TEMPLATE/question.yml:34]:
```suggestion
        - MacOS
```

File: .github/ISSUE_TEMPLATE/question.yml [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+labels: ["question"]
+body:
+  - type: markdown
+    attributes:
+      value: |
+        # Asking a question
+        If you have a question about PcapPlusPlus, please check the following resources before opening a new issue:
+        - [PcapPlusPlus documentation](https://pcapplusplus.github.io/docs/quickstart)
+        - [PcapPlusPlus API reference](https://pcapplusplus.github.io/docs/api)
+        - [GIthub issues tracker](https://github.com/seladb/PcapPlusPlus/issues)
```
Comment (clementperon) [inline .github/ISSUE_TEMPLATE/question.yml:12]:
```suggestion
        - [Github issues tracker](https://github.com/seladb/PcapPlusPlus/issues)
```

File: .github/ISSUE_TEMPLATE/bug.yml [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+      placeholder: ex. Build with commit hash 1234567
+    validations:
+      required: false
+  - type: dropdown
+    attributes:
+      label: "Operating systems tested on:"
+      multiple: true
+      options:
+        - Linux
+        - macOS
```
Comment (clementperon) [inline .github/ISSUE_TEMPLATE/bug.yml:69]:
```suggestion
        - MacOS
```

File: .github/ISSUE_TEMPLATE/bug.yml [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+        - FreeBSD
+        - Android
+        - Other (specify in the next question)
+    validations:
+      required: true
+  - type: input
+    attributes:
+      label: "Other operation systems (if applicable):"
+      description: If you select "Other", please specify the operation system you tested on.
+      placeholder: ex. ChromeOS, Solaris
```
Comment (clementperon) [inline .github/ISSUE_TEMPLATE/bug.yml:80]: We could also add the architecture if it's ARM64 or AMD64

**Review by clementperon — APPROVED** [review]

File: .github/ISSUE_TEMPLATE/question.yml (hunk 10 above)
Comment (tigercosmos) [reply .github/ISSUE_TEMPLATE/question.yml:34]: macOS is the correct spelling, please see https://en.wikipedia.org/wiki/MacOS

File: .github/ISSUE_TEMPLATE/bug.yml (hunk 13 above)
Comment (tigercosmos) [reply .github/ISSUE_TEMPLATE/bug.yml:80]: I am wondering was there any bug related to the architecture before? so far I didn't see any?

File: .github/ISSUE_TEMPLATE/question.yml (hunk 11 above)
Comment (tigercosmos) [reply .github/ISSUE_TEMPLATE/question.yml:12]: Thanks, let me correct to "GitHub", which is the correct one.

**Comment by tigercosmos** [comment]
@seladb   all issues addressed. please have a look.

**Comment by seladb** [comment]
@tigercosmos I believe only this one is left: https://github.com/seladb/PcapPlusPlus/pull/1413#discussion_r1620296642

File: .github/ISSUE_TEMPLATE/bug.yml [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+        ```
+
+        **Expected behavior**
+        A clear and concise description of what you expected to happen.
+
+    validations:
+      required: true
+  - type: dropdown
+    attributes:
+      label: "PcapPlusPlus versions tested on:"
```
[the remark this answers is not in the document]
Comment (tigercosmos) [reply .github/ISSUE_TEMPLATE/bug.yml:44]: @seladb  I followed the CPython style, but sure, let me change it.

[the remark this answers is not in the document]
Comment (tigercosmos) [reply .github/ISSUE_TEMPLATE/bug.yml:44]: fixed.

**Comment by tigercosmos** [comment]
@seladb  should all clear.

**Comment by seladb** [comment]
> @seladb should all clear.

@tigercosmos can you merge this branch to your fork's `master` branch so I can review it one last time?

**Comment by tigercosmos** [comment]
@seladb here you go.

> > @seladb should all clear.
> 
> @tigercosmos can you merge this branch to your fork's `master` branch so I can review it one last time?

**Review by seladb — APPROVED** [review]

**Pull request merged**